File: go/node_tracker/node_tracker_client.go
package node_tracker

import (
	"context"
	pbNodeTracker "di_store/pb/node_tracker"
	"di_store/util"

	"github.com/opentracing/opentracing-go"
	ot_log "github.com/opentracing/opentracing-go/log"
	"github.com/pkg/errors"
	"google.golang.org/grpc"
)

type NodeTrackerClient struct {
	rpcConn   *grpc.ClientConn
	rpcClient pbNodeTracker.NodeTrackerClient
}

func NewNodeTrackerClient(ctx context.Context, nodeTracker string) (*NodeTrackerClient, error) {
	span, ctx := opentracing.StartSpanFromContext(ctx, "NewNodeTrackerClient.Dial")
	defer span.Finish()
	conn, err := grpc.DialContext(ctx, nodeTracker, util.GrpcDialOption()...)
	if err != nil {
		span.LogFields(ot_log.Error(err))
		return nil, errors.Wrapf(err, "dial node tracker: %v", nodeTracker)
	}
	client := pbNodeTracker.NewNodeTrackerClient(conn)
	return &NodeTrackerClient{rpcConn: conn, rpcClient: client}, nil
}

func (client *NodeTrackerClient) RegisterStorageServer(
	hostname string, rpcPort int, plasmaSocket string, objTransferPort int, groupList []string) (
	*pbNodeTracker.RegisterStorageServerResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), util.CommonConfig.RequestTimeout)
	defer cancel()
	r, err := client.rpcClient.RegisterStorageServer(ctx, &pbNodeTracker.StorageServer{
		Hostname:        hostname,
		RpcPort:         int32(rpcPort),
		PlasmaSocket:    plasmaSocket,
		ObjTransferPort: int32(objTransferPort),
		GroupList:       groupList,
	})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.RegisterStorageServer")
	}
	return r, nil
}

func (client *NodeTrackerClient) UnregisterStorageServer(hostname string) (
	*pbNodeTracker.UnregisterStorageServerResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), util.CommonConfig.RequestTimeout)
	defer cancel()
	r, err := client.rpcClient.UnregisterStorageServer(ctx, &pbNodeTracker.StorageServer{
		Hostname: hostname,
	})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.UnregisterStorageServer")
	}
	return r, nil
}

func (client *NodeTrackerClient) RegisterStorageClient(serverHostname string) (*pbNodeTracker.RegisterStorageClientResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), util.CommonConfig.RequestTimeout)
	defer cancel()
	r, err := client.rpcClient.RegisterStorageClient(ctx, &pbNodeTracker.StorageClient{ServerHostname: serverHostname})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.RegisterStorageClient")
	}
	return r, nil
}

func (client *NodeTrackerClient) RegisterObject(ctx context.Context, objID, server string) (*pbNodeTracker.RegisterObjectResponse, error) {
	span, ctx := opentracing.StartSpanFromContext(ctx, "NodeTrackerClient.RegisterObject")
	defer span.Finish()

	ctx, cancel := context.WithTimeout(ctx, util.CommonConfig.RequestTimeout)
	defer cancel()
	r, err := client.rpcClient.RegisterObject(ctx, &pbNodeTracker.RegisterObjectRequest{ObjectIdHex: objID, ServerHostname: server})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.RegisterObject")
	}
	return r, nil
}

func (client *NodeTrackerClient) ServerInfo(ctx context.Context, serverList ...string) ([]*pbNodeTracker.StorageServer, error) {
	ctx, cancel := context.WithTimeout(ctx, util.CommonConfig.RequestTimeout)
	defer cancel()
	r, err := client.rpcClient.ServerInfo(ctx, &pbNodeTracker.ServerInfoRequest{ServerHostnameList: serverList})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.ServerInfo")
	}
	return r.StorageServerList, nil
}

func (client *NodeTrackerClient) ObjectInfo(ctx context.Context, objList ...string) ([]*pbNodeTracker.ObjectInfo, error) {
	span, ctx := opentracing.StartSpanFromContext(ctx, "NodeTrackerClient.ObjectInfo")
	defer span.Finish()

	r, err := client.rpcClient.ObjectInfo(ctx, &pbNodeTracker.ObjectInfoRequest{ObjectIdHexList: objList})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.ObjectInfo")
	}
	return r.ObjectInfoList, nil
}

func (client *NodeTrackerClient) ObjectDelete(objList ...string) (*pbNodeTracker.ObjectDeleteResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), util.CommonConfig.RequestTimeout)
	defer cancel()
	r, err := client.rpcClient.ObjectDelete(ctx, &pbNodeTracker.ObjectDeleteRequest{ObjectIdHexList: objList})
	if err != nil {
		return nil, errors.Wrapf(err, "NodeTrackerClient.ObjectDelete")
	}
	return r, nil
}

File: di_store/node_tracker/node_tracker_client.py
import grpc

from . import node_tracker_pb2
from . import node_tracker_pb2_grpc
from ..tracing import trace, wrap_channel

# todo grpc with multiprocessing
# https://github.com/grpc/grpc/issues/18321


@trace
class NodeTrackerClient:
    def __init__(self, node_tracker_host='127.0.0.1', node_tracker_port=50050):
        self.node_tracker_rpc_target = f'{node_tracker_host}:{node_tracker_port}'

        self.closed = False

    def close(self):
        if not self.closed:
            self.closed = True
            self.channel.close()

    def register_storage_server(self,
                                hostname,
                                ip_addr='', rpc_port=50051,
                                plasma_socket='/tmp/plasma'):
        assert hostname is not None
        with grpc.insecure_channel(self.node_tracker_rpc_target) as channel:
            stub = node_tracker_pb2_grpc.NodeTrackerStub(channel)
            response = stub.register_storage_server(
                node_tracker_pb2.StorageServer(
                    hostname=hostname,
                    ip_addr=ip_addr,
                    rpc_port=rpc_port,
                    plasma_socket=plasma_socket
                ))
            return response

    def register_storage_client(self, server_hostname):
        assert server_hostname is not None
        with grpc.insecure_channel(self.node_tracker_rpc_target) as channel:
            stub = node_tracker_pb2_grpc.NodeTrackerStub(wrap_channel(channel))
            response = stub.register_storage_client(
                node_tracker_pb2.StorageClient(
                    server_hostname=server_hostname
                ))
            return response

    def register_object(self, object_id_hex, server_hostname, push_hostname_list, push_group_list):
        with grpc.insecure_channel(self.node_tracker_rpc_target) as channel:
            stub = node_tracker_pb2_grpc.NodeTrackerStub(wrap_channel(channel))
            response = stub.register_object(
                node_tracker_pb2.RegisterObjectRequest(
                    object_id_hex=object_id_hex,
                    server_hostname=server_hostname,
                    push_hostname_list=push_hostname_list,
                    push_group_list=push_group_list
                ))
            return response

    def server_info(self, server_hostname_list=None):
        request = node_tracker_pb2.ServerInfoRequest()
        if server_hostname_list:
            if isinstance(server_hostname_list, list):
                request.server_hostname_list.extend(server_hostname_list)
            else:
                request.server_hostname_list.append(server_hostname_list)
        with grpc.insecure_channel(self.node_tracker_rpc_target) as channel:
            stub = node_tracker_pb2_grpc.NodeTrackerStub(wrap_channel(channel))
            response = stub.server_info(request)
            return response.storage_server_list

    def object_info(self, object_id_hex_list):
        request = node_tracker_pb2.ObjectInfoRequest()
        if isinstance(object_id_hex_list, list):
            request.object_id_hex_list.extend(object_id_hex_list)
        else:
            request.object_id_hex_list.append(object_id_hex_list)

        with grpc.insecure_channel(self.node_tracker_rpc_target) as channel:
            stub = node_tracker_pb2_grpc.NodeTrackerStub(wrap_channel(channel))
            response = stub.object_info(request)
            return response.object_info_list

    def object_delelte(self, object_id_hex_list):
        request = node_tracker_pb2.ObjectDeleteRequest()
        if isinstance(object_id_hex_list, list):
            request.object_id_hex_list.extend(object_id_hex_list)
        else:
            request.object_id_hex_list.append(object_id_hex_list)

        with grpc.insecure_channel(self.node_tracker_rpc_target) as channel:
            stub = node_tracker_pb2_grpc.NodeTrackerStub(wrap_channel(channel))
            return stub.object_delete(request)
